src/index.js

```diff
@@ -458,11 +459,23 @@ class Libp2p extends EventEmitter {
         }
 
         discoveryService.on('peer', this._onDiscoveryPeer)
-        this._discovery.push(discoveryService)
+        this._discovery.set(DiscoveryService.tag, discoveryService)
+      }
+    }
+
+    // Discovery modules
+    for (const DiscoveryService of this._modules.peerDiscovery || []) {
+      setupService(DiscoveryService)
+    }
+
+    // Transport modules with discovery
+    for (const Transport of this.transportManager.getTransports()) {
+      if (Transport.discovery) {
+        setupService(Transport.discovery)
       }
     }
 
-    return this._discovery.map(d => d.start())
+    return Array.from(this._discovery.values()).map(d => d.start())
```

Comment:
We should promisify these calls since `start` is supposed to be async, right now it's just an array of promises. The jsdocs already mention yielding a promise, so this will actually do that. Additionally, you can just pass the map function to `Array.from`, which should avoid double iteration.
```suggestion
    return Promise.all(Array.from(this._discovery.values(), d => d.start()))
```